sbin/anti-evil-maid-lib-tpm1

```diff
@@ -0,0 +1,220 @@
+
+updatefreshness() {
+    # takes a path to the new freshness token as an argument and
+    # stores its sha1 hash in the appropriate freshness token slot
+    # of the TPM NVRAM area; second argument is the AEM boot device
+    # label suffix
+    if [ ! -e "$TPM_FRESHNESS_PASSWORD_FILE" ]; then
+        message "TPM NVRAM area AUTHWRITE password file does not exist!"
+        return 1
+    fi
+
+    if ! _slot=$(suffixtoslot "$2"); then
+        message "Suffix '$2' not in DB, attempting to create..."
+        if ! _slot=$(assignslottosuffix "$2"); then
+            message "Failed to add suffix '$2' into DB!"
+            return 1
+        fi
+    fi
+
+    _pw=$(cat "$TPM_FRESHNESS_PASSWORD_FILE")
+    hashfile "$1" | unhex \
+    | tpm_nvwrite_stdin -i "$TPM_FRESHNESS_INDEX" \
+      -n "$((_slot * 20))" -s 20 --password="$_pw"
+}
+
+revokefreshness() {
+    # invalidates the freshness token of a specified AEM media (by its
+    # label suffix
+    _suff=$1
+    if _slot=$(suffixtoslot "$_suff"); then
+        message "Revoking freshness token for AEM media w/ suffix '$_suff'..."
+        _pw=$(cat "$TPM_FRESHNESS_PASSWORD_FILE")
+        if tpm_nvwrite -i "$TPM_FRESHNESS_INDEX" \
+                -n "$((_slot * 20))" -s 20 \
+                --password="$_pw" -m "0xff"; then
+            message "Done."
+        else
+            message "Failed!"
+        fi
+    else
+        message "AEM device with label suffix '$_suff' not found in DB!"
+    fi
+}
+resetfreshness() {
+    # invalidates ALL freshness tokens
+    message "Invalidating **ALL** freshness tokens..."
+    _pw=$(cat "$TPM_FRESHNESS_PASSWORD_FILE")
+    if tpm_nvwrite -i "$TPM_FRESHNESS_INDEX" \
+            -s "$((TPM_FRESHNESS_SLOTS * 20))" \
+            --password="$_pw" -m "0xff"; then
+        message "Done."
+    else
+        message "Failed!"
+    fi
+}
+
+destroytpmnvram() {
+    # releases the TPM NVRAM area; TPM owner pw as first argument
+    tpm_nvrelease -i "$TPM_FRESHNESS_INDEX" --pwdo="$1"
+}
+
+listbadpcrs() {
+    # prints those standard PCRs configured to be used via $SEAL which haven't
+    # been extended, output is empty there are no such PCRs
+    _pcrs=$(printf %s "$SEAL" | grep -Eo '\b1[3789]\b') || true
+    grep -E "^PCR-(${_pcrs//$'\n'/|}):( 00| FF){20}" "$SYSFS_TPM_DIR"/pcrs
+}
+
+tpmowned() {
+    # checks whether TPM is already owned, signals results with exit code
+    [ "$(cat "$SYSFS_TPM_DIR"/owned)" -ne 0 ]
+}
+
+provisiontpmid() {
+    # stores TPM ID into an NVRAM entry 
+    _tpm_id_index=$(tpm_id -i)
+    _opw=$(cat "$TPM_OWNER_PASSWORD_FILE")
+    # create a write-once NVRAM area
+    tpm_nvdefine -i "$_tpm_id_index" -s 20 -p "WRITEDEFINE|WRITEALL" \
+                 --pwdo="$_opw"
+    # generate a random ID and write it into NVRAM
+    head -c 20 /dev/random | tpm_nvwrite_stdin -i "$_tpm_id_index" -s 20
+    # lock the area to prevent non-owners from changing ID
+    tpm_nvwrite -i "$_tpm_id_index" -s 0
+}
+
+postprovisioning() {
```

Comment:
Is this specific to the TPM family? I haven't seen any significant changes touching `$TPMS_DIR` in TPM2.0 code, but maybe I missed it.

Reply:
I see.